### DISCUSSION TITLE
Resolving `relationship will copy column ... which conflicts with`

### DISCUSSION THREAD

**Discussion by Cehovoi** [sqlalchemy/sqlalchemy#10037] [Usage Questions] [answered]
### Describe the bug

In our legacy code base, after upgrading to sqlalchemy from 1.3.17 to 2.0.15 we received many warnings about relationships will copy column ... which conflicts with...
Overlaps argument and view only=True does not resolve warnings.
After examining the documentation, stackoverflow and github, I came to the conclusion that in our case there is no overlap. In our case we use some constants as additional keys, but sqlalchemy doesn't take that into account.
How can I disable all warnings?

### SQLAlchemy Version in Use

2.0.15

### DBAPI (i.e. the database driver)

oracledb

### Database Vendor and Major Version

oracle 19.8

### Python Version

3.10

### Operating system

MacOs

### To Reproduce

```python
class ObjectTypesConstant(Enum):
    Country = 1
    Region = 2
    Operator = 33
    Territory = 57
    Branch = 6
    SubBranch = 58
    User = 19
    Partner = 3
    Salepoint = 5



class ObjectParam(Model):

    object_param_id = sa.Column(sa.Integer, primary_key=True, name='op_id', autoincrement=True)
    obj_id = sa.Column(sa.Integer, nullable=False, name='op_obj_id')
    obj_type = sa.Column(sa.Integer, nullable=False, name='op_obj_type')


class Country(Model):

    pk = sa.Column(sa.Integer, primary_key=True, name='c_id', autoincrement=True)
    type_code = ObjectTypesConstant.Country.value

    params = relationship("ObjectParam",
                          primaryjoin=f"and_(ObjectParam.obj_id == Country.pk, {type_code}" +
                                      " == ObjectParam.obj_type)",
                          foreign_keys='ObjectParam.obj_id', lazy="dynamic")


class SubBranch(Model):
    pk = sa.Column(sa.Integer, primary_key=True, name='sbr_id')
    type_code = ObjectTypesConstant.SubBranch.value
    params = relationship("ObjectParam",
                          primaryjoin=f"and_(ObjectParam.obj_id == SubBranch.pk, {type_code}" +
                                      " == ObjectParam.obj_type)",
                          foreign_keys='ObjectParam.obj_id', lazy="dynamic")
```

### Error

SAWarning: relationship 'SubBranch.params' will copy column skad_subbranch.sbr_id to column skad_object_param.op_obj_id, which conflicts with relationship(s): 'Country.params' (copies skad_country.c_id to skad_object_param.op_obj_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="params"' to the 'SubBranch.params' relationship. (Background on this warning at: https://sqlalche.me/e/20/qzyx) (This warning originated from the `configure_mappers()` process, which was invoked automatically in response to a user-initiated operation.)

**Comment by zzzeek** [comment]
> ### Describe the bug
> 
> In our legacy code base, after upgrading to sqlalchemy from 1.3.17 to 2.0.15 we received many warnings about relationships will copy column ... which conflicts with... Overlaps argument and view only=True does not resolve warnings. After examining the documentation, stackoverflow and github, I came to the conclusion that in our case there is no overlap. In our case we use some constants as additional keys, but sqlalchemy doesn't take that into account.

that's correct, as the message indicates, SQLAlchemy only sees the columns that are being copied and would not know how the additional constant changes that in practice.

> How can I disable all warnings?

the warning contains the exact code that you would use:

> To silence this warning, add the parameter 'overlaps="params"' to the 'SubBranch.params' relationship. 

just as it says:

```py
class SubBranch(Model):
    # ...
    params = relationship(..., overlaps="params")
```

**Reply by Cehovoi** [reply]
It is not working, when i add overlaps="params", the warning still remains.

**Reply by zzzeek** [reply, marked as answer]
can't reproduce.  Please run this demonstration case, which here illustrates the warning appearing only when SHOW_WARNING=True.      Ensure that for each warning emitted, the instructions in the warning are written out in code, e.g. 

> To silence this warning, add the parameter 'overlaps="params"' to the 'SubBranch.params' relationship.

that means locate the `SubBranch` class, locate the `params` relationship and add "overlaps="params"".     if you then have say **another** warning, such as :

> To silence this warning, add the parameter 'overlaps="xyz"' to the 'SomeOtherClass.xyz' relationship.

then you have to locate the `xyz` relationship on `SomeOtherClass`, etc.

Please run this program below and see what results you get.   Send me a modified version that illustrates the warning not being cleared.

```py
from enum import Enum

import sqlalchemy as sa
from sqlalchemy.orm import configure_mappers
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship


class ObjectTypesConstant(Enum):
    Country = 1
    Region = 2
    Operator = 33
    Territory = 57
    Branch = 6
    SubBranch = 58
    User = 19
    Partner = 3
    Salepoint = 5


SHOW_WARNING = False

Model = declarative_base()


class ObjectParam(Model):
    __tablename__ = "object_param"
    object_param_id = sa.Column(
        sa.Integer, primary_key=True, name="op_id", autoincrement=True
    )
    obj_id = sa.Column(sa.Integer, nullable=False, name="op_obj_id")
    obj_type = sa.Column(sa.Integer, nullable=False, name="op_obj_type")


class Country(Model):
    __tablename__ = "country"

    pk = sa.Column(
        sa.Integer, primary_key=True, name="c_id", autoincrement=True
    )
    type_code = ObjectTypesConstant.Country.value

    params = relationship(
        "ObjectParam",
        primaryjoin=f"and_(ObjectParam.obj_id == Country.pk, {type_code}"
        + " == ObjectParam.obj_type)",
        foreign_keys="ObjectParam.obj_id",
        lazy="dynamic",
    )


class SubBranch(Model):
    __tablename__ = "subbranch"
    pk = sa.Column(sa.Integer, primary_key=True, name="sbr_id")
    type_code = ObjectTypesConstant.SubBranch.value

    if SHOW_WARNING:
        params = relationship(
            "ObjectParam",
            primaryjoin=f"and_(ObjectParam.obj_id == SubBranch.pk, {type_code}"
            + " == ObjectParam.obj_type)",
            foreign_keys="ObjectParam.obj_id",
            lazy="dynamic",
        )
    else:
        params = relationship(
            "ObjectParam",
            primaryjoin=f"and_(ObjectParam.obj_id == SubBranch.pk, {type_code}"
            + " == ObjectParam.obj_type)",
            foreign_keys="ObjectParam.obj_id",
            lazy="dynamic",
            overlaps="params",
        )


configure_mappers()

```

**Comment by Cehovoi** [comment]
Thank you so much!  Run configure_mappers() resolve my problem.
In my case there are more than 70 places with overlap, and I ran my parser to parse the warning list and insert the ovelaps argument into the code. With configure_mappers() the warnings are gone.
